src/playwright/pages/base.page.ts

```diff
@@ -0,0 +1,103 @@
+  | 'Accessibility Statement';
+type FooterLink = AccessibilityStatementLink;
+
+type SearchButton = 'Hae tapahtumia' | 'Sök evenemang' | 'Search events';
+type TargetGroupButton = 'Kohderyhmät' | 'Målgrupp' | 'Target groups';
+
+type ZeroToTwoYears = '0-2 vuotiaat' | '0-2 år' | '0-2 years';
+type TargetGroupOption = ZeroToTwoYears;
+type MainContentOption = TargetGroupOption;
+
+type MainContentTextBox = SearchButton;
+
+type MainContentButton = SearchButton | TargetGroupButton;
+
+export class BasePage {
+  protected readonly page: Page;
+  protected readonly h1: Locator;
+  protected readonly footer: Locator;
+  protected readonly header: Locator;
+  protected readonly mainContent: Locator;
+  protected readonly cookieConsentBox: Locator;
+  protected readonly cookieConsentApprove: Locator;
+
+  constructor(page: Page) {
+    if (this.constructor === BasePage) {
+      throw new Error(
+        'Abstract BasePage class cannot be instantiated directly'
+      );
+    }
+    this.page = page;
+    this.h1 = page.locator('h1');
+    this.header = page.getByRole('banner');
+    this.footer = page.getByRole('contentinfo');
+    this.mainContent = page.locator('#main-content');
+    this.cookieConsentBox = page.locator('#cookie-consent-content');
+    this.cookieConsentApprove = page.getByTestId(
+      'cookie-consent-approve-required-button'
+    );
+  }
+
+  async isReady() {
+    await this.page.waitForLoadState('networkidle', IN_20s); // Network can be very slow
+  }
+
+  async reload() {
+    return await this.page.reload();
+  }
+
+  async clickFooterLink(name: FooterLink) {
+    await this.footer.getByRole('link', { name }).first().click();
+  }
+
+  async clickHeaderButton(name: HeaderButton) {
+    await this.header.getByRole('button', { name }).first().click();
+  }
+
+  async clickSearchOption(name: MainContentOption) {
+    await this.mainContent.getByRole('option', { name }).first().click();
+  }
+
+  async fillSearchTextBox(name: MainContentTextBox, value: string) {
+    const searchTextBox = this.page.getByRole('textbox', { name });
+    await searchTextBox.click(); // First open the text box by clicking on it
+    await searchTextBox.fill(value);
+  }
+
+  async clickSearchButton(name: MainContentButton) {
+    await this.mainContent.getByRole('button', { name }).first().click();
+  }
+
+  async isCookieConsentBoxVisible(timeout?: number) {
+    await expect(this.cookieConsentBox).toBeVisible({ timeout });
+  }
+
+  async isCookieConsentBoxHidden(timeout?: number) {
+    await expect(this.cookieConsentBox).not.toBeVisible({ timeout });
+  }
+
+  async isCookieConsentBoxHiddenAfterMs(delayMs: number) {
+    await this.page.waitForTimeout(delayMs);
+    await this.isCookieConsentBoxHidden(0); // Instant check
+  }
+
+  async approveCookies() {
+    await this.cookieConsentApprove.click();
+  }
+
+  protected async isTitle(text: string, timeout?: Timeout) {
+    await expect(this.h1).toContainText(text, timeout);
+  }
```

Comment:
Every page should have a title, so this function is OK to have in base page, but I think it should be renamed to `hasTitle`.

```suggestion
  protected async hasTitle(text: string, timeout?: Timeout) {
    await expect(this.h1).toContainText(text, timeout);
  }
```

Header 1 should always be rendered e.g. for accessibility reasons. Should we also test a title-tag in head meta section?

Reply:
Head's meta section doesn't always have the same value as h1, e.g. in production environment:

Finnish front page:
```html
<head>
<title>Kultus</title>
<meta property="og:description" content="Oppimisen elämyksiä koko kaupungissa! Taidetta, kulttuuria ja tapahtumia päiväkodeille ja kouluille.">
<meta property="og:title" content="Kultus">
</head>
```

Swedish front page:
```html
<head>
<title>Kultus</title>
<meta name="description" content="Upplevelser av att lära sig runt om i stan! Konst, kultur och evenemang för skola och småbarnspedagogik.">
<meta property="og:title" content="Kultus">
</head>
```

English front page:
```html
<head>
<title>Kultus</title>
<meta name="description" content="Learning experiences around the city! Arts, culture, and events for daycare and schools.">
<meta property="og:title" content="Kultus">
</head>
```

And for e.g. the accessibility statement CMS page in production environment:

Finnish accessibility statement CMS page:
```html
<head>
<meta property="og:title" content="Saavutettavuusseloste">
<meta name="twitter:title" content="Saavutettavuusseloste">
<meta property="og:description" content="T&amp;auml;m&amp;auml; saavutettavuusseloste koskee Helsingin kaupungin Kultus &amp;#x2d;verkkosivustoa. Sivuston osoite on https://kultus.hel.fi/fi ja https://kultus&amp;#x2d;admin.hel.fi/fi T&amp;auml;m&amp;auml; sivusto on julkaistu 05.10.2020 j&amp;auml;lkeen.">
</head>
```
but it didn't seem to have a <title> tag.

So, because the data between h1 and <head><title> or <head><meta> tags doesn't always match exactly (e.g. on the front page it doesn't, on the Finnish accessibility statement CMS page it did match with meta og:title) it makes testing both head info and h1 with the same value impractical at least in the general case.

Of course we could check for <head><title> or <head><meta property="og:title"> separately, but then we'd have to probably somehow change the function namings so it'd be understandable as to what is actually done.

Reply:
Lets just go with the visible Heading 1 only at this point.